### PR TITLE
partially automate extraction of concepts.csv

### PR DESCRIPTION
**Pull request by efx**
I cobbled together my knowledge of shell and `nu`.

1. save the following to `_test/migrate-concept-csv.sh`:
```sh

echo "uuid,slug,name,blurb"
for slug in $(cut -d, -f1 < reference/concepts.csv | sort | tr . -); do
    # skip concepts that have a directory
    if [ -d "concepts/$slug" ]; then
        continue
    fi
    # create a human readable name
    name=$(echo "$slug" | nu --stdin -c "str find-replace -a '-' ' ' | str capitalize")
    echo "$(./bin/configlet uuid),$slug,$name,"
done
```
2. run the below to generate formatted JSON from it for manual pasting into
`config.json`:
```
./_test/migrate-concept-csv.sh  | from csv | to json | ^jq --indent 4 | clip
```
3. manually remove `vec` from that list.
4. Save below to `_test/migrate-concept-todos.sh` and run it:
```sh

config=./config.json
for slug in $(jq --raw-output '.concepts[] |.slug' $config); do
    # skip concepts that have a directory
    if [ -d "concepts/$slug" ]; then
        continue
    fi
    concept_dir="concepts/$slug"
    mkdir -p "$concept_dir"
    echo "TODO" > "$concept_dir/about.md"
    echo "TODO" > "$concept_dir/introduction.md"
    echo "[]" > "$concept_dir/links.json"
done
```

The extraction needs to happen before #1123.

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -1378,6 +1378,468 @@
...
+        "slug": "collect",
+        "name": "Collect",
+        "blurb": ""
+    },
+    {
+        "uuid": "daf92b42-f82f-4a9f-a709-a39e74843891",
+        "slug": "concept",
+        "name": "Concept",
+        "blurb": ""
+    },
```
Comment (coriolinus) [inline config.json:1464]:
```suggestion
```

File: config.json [hunk 2]
```diff
@@ -1378,6 +1378,468 @@
...
+    {
+        "uuid": "4e5ca742-379c-49da-9f57-54a740f3e632",
+        "slug": "conditionals",
+        "name": "Conditionals",
+        "blurb": ""
+    },
+    {
+        "uuid": "44644a80-8b1f-4a52-a999-97a3f5f9db6b",
+        "slug": "const-and-static",
+        "name": "Const and static",
```
Comment (coriolinus) [inline config.json:1474]:
```suggestion
        "name": "Const and Static",
```

File: config.json [hunk 3]
```diff
@@ -1378,6 +1378,468 @@
...
+        "blurb": ""
+    },
+    {
+        "uuid": "992f5e5d-72a8-4f52-95e1-3bbe6c4f4689",
+        "slug": "count-and-sum",
+        "name": "Count and sum",
```
Comment (coriolinus) [inline config.json:1480]:
```suggestion
        "name": "Count and Sum",
```

File: config.json [hunk 4]
```diff
@@ -1378,6 +1378,468 @@
...
+    {
+        "uuid": "445f2f63-e5fb-4500-93b3-19afac753bfa",
+        "slug": "format-macro",
+        "name": "Format macro",
+        "blurb": ""
+    },
+    {
+        "uuid": "44d3d261-443d-4bbc-834e-5d35aca0cab8",
+        "slug": "from-into-iterator",
+        "name": "From into iterator",
```
Comment (coriolinus) [inline config.json:1540]:
```suggestion
        "name": "FromIterator, IntoIterator, and .collect()",
```

File: config.json [hunk 5]
```diff
@@ -1378,6 +1378,468 @@
...
+        "slug": "functions",
+        "name": "Functions",
+        "blurb": ""
+    },
+    {
+        "uuid": "3848f069-afe9-4882-95ba-4112b565bdab",
+        "slug": "functions-anonymous",
+        "name": "Functions anonymous",
+        "blurb": ""
+    },
```
Comment (coriolinus) [inline config.json:1554]:
```suggestion
```

Anonymous functions are only possible via closures in Rust, so there's no point in having distinct concepts.

File: config.json [hunk 6]
```diff
@@ -1378,6 +1378,468 @@
...
+    {
+        "uuid": "bd7d1ccb-feac-482d-86bd-3d611172bb3f",
+        "slug": "functions-closures",
+        "name": "Functions closures",
```
Comment (coriolinus) [inline config.json:1558]:
```suggestion
        "slug": "closures",
        "name": "Closures",
```

File: config.json [hunk 7]
```diff
@@ -1378,6 +1378,468 @@
...
+        "blurb": ""
+    },
+    {
+        "uuid": "99e5f5e4-c202-4ad2-917f-068ff56e1c03",
+        "slug": "functions-higher-order",
+        "name": "Functions higher order",
```
Comment (coriolinus) [inline config.json:1564]:
```suggestion
        "slug": "higher-order-functions",
        "name": "Higher order functions",
```

File: config.json [hunk 8]
```diff
@@ -1378,6 +1378,468 @@
...
+        "slug": "futures",
+        "name": "Futures",
+        "blurb": ""
+    },
+    {
+        "uuid": "9cb5cf22-eea4-4434-bfe9-1d05fe883c7b",
+        "slug": "generic",
+        "name": "Generic",
+        "blurb": ""
+    },
```
Comment (coriolinus) [inline config.json:1578]:
```suggestion
```

Dupe: "generics"

File: config.json [hunk 9]
```diff
@@ -1378,6 +1378,468 @@
...
+    {
+        "uuid": "fc7e0724-11de-4471-95d1-4cd76324bebe",
+        "slug": "generics",
+        "name": "Generics",
+        "blurb": ""
+    },
+    {
+        "uuid": "aab1a247-8c64-48fd-bf76-3384957aadd6",
+        "slug": "hashmap",
+        "name": "Hashmap",
```
Comment (coriolinus) [inline config.json:1594]:
```suggestion
        "name": "HashMap",
```

File: config.json [hunk 10]
```diff
@@ -1378,6 +1378,468 @@
...
+        "blurb": ""
+    },
+    {
+        "uuid": "8eaa0431-2743-4ff0-a7ba-2b2ba496fb92",
+        "slug": "hashset",
+        "name": "Hashset",
```
Comment (coriolinus) [inline config.json:1600]:
```suggestion
        "name": "HashSet",
```

File: config.json [hunk 11]
```diff
@@ -1378,6 +1378,468 @@
...
+        "blurb": ""
+    },
+    {
+        "uuid": "dc958fc2-e1f6-4ff9-8f9c-9bde565d8ec9",
+        "slug": "if-while-let",
+        "name": "If while let",
```
Comment (coriolinus) [inline config.json:1606]:
```suggestion
        "name": "`if let` and `while let`",
```

File: config.json [hunk 12]
```diff
@@ -1378,6 +1378,468 @@
...
+    {
+        "uuid": "952406cb-dcb1-47df-8129-b996b32a6cc4",
+        "slug": "loops",
+        "name": "Loops",
+        "blurb": ""
+    },
+    {
+        "uuid": "f16a9f91-7fbd-406a-9642-5cb5f9f52590",
+        "slug": "macros-declarative",
+        "name": "Macros declarative",
```
Comment (coriolinus) [inline config.json:1648]:
```suggestion
        "name": "Declarative Macros",
```

File: config.json [hunk 13]
```diff
@@ -1378,6 +1378,468 @@
...
+        "blurb": ""
+    },
+    {
+        "uuid": "9e7b6f00-c30d-4a44-934f-e762203ebc2f",
+        "slug": "macros-procedural",
+        "name": "Macros procedural",
```
Comment (coriolinus) [inline config.json:1654]:
```suggestion
        "name": "Procedural Macros",
```

File: config.json [hunk 14]
```diff
@@ -1378,6 +1378,468 @@
...
+    {
+        "uuid": "5a1896c8-b479-4bc4-89e7-a74d34056bde",
+        "slug": "result",
+        "name": "Result",
+        "blurb": ""
+    },
+    {
+        "uuid": "5173a1e4-5909-4dd3-a348-1750c096f1de",
+        "slug": "rwlock",
+        "name": "Rwlock",
```
Comment (coriolinus) [inline config.json:1744]:
```suggestion
        "name": "RwLock",
```

File: config.json [hunk 15]
```diff
@@ -1378,6 +1378,468 @@
...
+    {
+        "uuid": "ac37efdf-a1c5-4526-bd7a-fe1c4cba1656",
+        "slug": "static-lifetime",
+        "name": "Static lifetime",
+        "blurb": ""
+    },
+    {
+        "uuid": "07a32f06-1268-4f17-abbe-7b9c7fcc7995",
+        "slug": "std-thread",
+        "name": "Std thread",
```
Comment (coriolinus) [inline config.json:1774]:
```suggestion
        "name": "std::thread",
```

File: config.json [hunk 16]
```diff
@@ -1378,6 +1378,468 @@
...
+        "slug": "string",
+        "name": "String",
+        "blurb": ""
+    },
+    {
+        "uuid": "9fe8a452-a566-4910-807b-6a0fb69265ce",
+        "slug": "string-vs-str",
+        "name": "String vs str",
+        "blurb": ""
+    },
```
Comment (coriolinus) [inline config.json:1788]:
```suggestion
```

Pretty sure we already have those concepts implemented.

File: config.json [hunk 17]
```diff
@@ -1378,6 +1378,468 @@
...
+        "slug": "structs-tuple",
+        "name": "Structs tuple",
+        "blurb": ""
+    },
+    {
+        "uuid": "46eeb655-ab71-4c04-bf6c-f87fd9753001",
+        "slug": "structs-zero-size",
+        "name": "Structs zero size",
+        "blurb": ""
+    },
```
Comment (coriolinus) [inline config.json:1806]:
```suggestion
```

More dupes of already-existing concepts.

**Review by coriolinus — APPROVED** [review]
This is great! Did a little bit of manual review just to clean up the low-hanging fruit a bit. Definitely stands to do with some more cleanup, but even as-is, this should unblock #1123.

File: config.json (hunk 16 above)
Comment (efx) [reply config.json:1788]: Looks like we have string-use but not string-vs-str. That said, I want to resolve #1057 so I will add it to `config.json` when I do that soonish!

**Pull request merged**